File: netlogo-gui/src/main/window/QuickHelp.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.window;

import org.nlogo.core.I18N;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.nio.file.Path;

final strictfp class QuickHelp {

  // not instantiable
  private QuickHelp() {
    throw new IllegalStateException();
  }

  /**
   * This action is to provide context-sensitive help in the procedures
   * tab.  Based on what their cursor is on, it should jump the user
   * to the proper documentation, found in docs/dict/
   * -- which is just a decomposed version of docs/dictionary.html
   * <p/>
   * Maybe it would be best to pull everything out of some new file,
   * maybe xml format, and we could autogenerate the primitives dictionary file
   * from that.
   * <p/>
   * ~Forrest (4/26/2007)
   */
  private static boolean quickHelpLoaded = false;
  private static final String QUICKHELPWORDS_PATH = "/system/dict.txt";
  private static final String QUICKHELPWORDS_PATH3D = "/system/dict3d.txt";
  private static Map<String, String> quickHelpWords;
  private static Map<String, String> quickHelpWords3d;

  private static Map<String, String> loadHelp(String path) {
    String[] lines = org.nlogo.api.FileIO$.MODULE$.getResourceAsStringArray(path);
    HashMap<String, String> words =
        new HashMap<String, String>();
    for (int i = 0; i < lines.length; i++) {
      String anchor = lines[i].substring(0, lines[i].indexOf(' '));
      String file = lines[i].substring(lines[i].indexOf(' ') + 1);
      words.put(anchor, file);
    }
    return words;
  }

  private static Path docPath(String docName) {
    return org.nlogo.swing.BrowserLauncher.docPath(docName);
  }

  private static void openDictionary(java.awt.Component comp, String theWord,
                                     Map<String, String> words) {
    String theFile = words.get(theWord);
    org.nlogo.swing.BrowserLauncher.openPath(comp, docPath("dict/" + theFile), null);
  }

  public static void doHelp(java.awt.Component comp, String token, boolean is3D) {
    if (!quickHelpLoaded) {
      quickHelpWords = loadHelp(QUICKHELPWORDS_PATH);
      quickHelpWords3d = loadHelp(QUICKHELPWORDS_PATH3D);
      quickHelpLoaded = true;
    }
    if (token == null) {
      return;
    }
    token = token.toLowerCase();
    // if there is an entry in the 3D dictionary then it overrides
    // the 2D entry ev 10/25/07
    if (is3D && quickHelpWords3d.containsKey(token)) {
      openDictionary(comp, token, quickHelpWords3d);
    } else if (quickHelpWords.containsKey(token)) {
      openDictionary(comp, token, quickHelpWords);
    } else {
      if (0 == javax.swing.JOptionPane.showConfirmDialog
          (comp, I18N.guiJ().getN("tabs.code.rightclick.quickhelp.notfound", token.toUpperCase()),
           I18N.guiJ().get("common.netlogo"), javax.swing.JOptionPane.YES_NO_OPTION)) {
        org.nlogo.swing.BrowserLauncher.openPath(comp, docPath("index2.html"), null);
      }
    }
  }
}

File: netlogo-gui/src/main/window/QuickHelp.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.window

import java.awt.Component
import java.net.URI
import java.nio.file.{ Files, Path, Paths }
import java.util.Locale

import org.nlogo.api.Version
import org.nlogo.core.I18N
import org.nlogo.swing.{ BrowserLauncher, OptionPane }

object QuickHelp {
  /**
   * This action is to provide context-sensitive help in the procedures
   * tab.  Based on what their cursor is on, it should jump the user
   * to the proper documentation, found in docs/dict/
   * -- which is just a decomposed version of docs/dictionary.html
   * <p/>
   * Maybe it would be best to pull everything out of some new file,
   * maybe xml format, and we could autogenerate the primitives dictionary file
   * from that.
   * <p/>
   * ~Forrest (4/26/2007)
   */

  private val docsRoot: String = System.getProperty("netlogo.docs.dir")

  // if we're not in 3D don't load the 3d dictionary words
  // cause we don't need 'em and they'll override the 2d
  // dictionary ev 10/25/07
  private lazy val quickHelpWords: Map[String, Entry] =
    loadHelp(false)

  private lazy val quickHelpWords3d: Map[String, Entry] = {
    if (Version.is3D) {
      loadHelp(true)
    } else {
      Map()
    }
  }

  def docPath(anchor: Option[String]): Path =
    Paths.get(docsRoot, "NetLogo_User_Manual.pdf" + anchor.fold("")("#" + _))

  private def loadHelp(threed: Boolean): Map[String, Entry] = {
    Files.readString(Paths.get(docsRoot, "manual-links.csv")).split('\n').flatMap { line =>
      val split = line.split(',')

      Entry.parse(split.tail, threed).map((split.head, _))
    }.toMap
  }

  private def openDictionary(comp: Component, word: String, words: Map[String, Entry]): Unit = {
    BrowserLauncher.tryOpenURI(
      comp,
      new URI(s"https://docs.netlogo.org/${Version.versionNumberNo3D}/${words(word).docsUrl}"),
      docPath(Some(words(word).pdfAnchor))
    )
  }

  def doHelp(comp: Component, token: String): Unit = {
    if (token == null)
      return

    val tokenLower = token.toLowerCase

    // if there is an entry in the 3D dictionary then it overrides
    // the 2D entry ev 10/25/07
    if (quickHelpWords3d.contains(tokenLower)) {
      openDictionary(comp, tokenLower, quickHelpWords3d)
    } else if (quickHelpWords.contains(tokenLower)) {
      openDictionary(comp, tokenLower, quickHelpWords)
    } else {
      if (new OptionPane(comp, I18N.gui.get("common.netlogo"),
                         I18N.gui.getN("tabs.code.rightclick.quickhelp.notfound",
                         tokenLower.toUpperCase(Locale.ENGLISH)), OptionPane.Options.OkCancel,
                         OptionPane.Icons.Error).getSelectedIndex == 0)
        BrowserLauncher.tryOpenURI(comp,
          new URI(s"https://docs.netlogo.org/${Version.versionNumberNo3D}/dictionary.html"),
          docPath(Some("dictionary-netlogo-dictionary")))
    }
  }

  private object Entry {
    def parse(els: Array[String], threed: Boolean): Option[Entry] = {
      if (els.size == 3 && (els(0) == "3d") == threed) {
        Some(Entry(els(1), els(2)))
      } else {
        None
      }
    }
  }

  private case class Entry(docsUrl: String, pdfAnchor: String)
}
